[WIP] Add error sync/handle logic for event manager

zimengyang opened the pull request:
Working in progress pr for event-manager error handling. 
* Sync with underlying k8s deployment for real status, and storing reason
* Put entity to ERROR status and try to recover.

zimengyang committed:
Add error sync/handle logic

kars7e commented on pkg/event-manager/drivers/k8s_errors.go:
```diff
@@ -0,0 +1,61 @@
+	errReasonUnknown                 = "Unknown"
+)
+
+var (
```
This seems ok, but I try to avoid using reflect if possible. It seems that what you are trying to achieve here is to have a unique string name for each error. You can't use Error() to return that because you are already returning the the nested error's `Error()`. How about one of these:
* You could define an interface, e.g. `Causer`, that has a `Cause() error` method. This method returns the underlying error. Then you can use `Error()` to return what you care about, and if you care about the underlying error, you can check `if ok := err.(Causer); ok`, and then use `Cause().Error()` to to get the underlying error message. In fact, something like that is used in [`pkg/errors`](https://github.com/pkg/errors/blob/816c9085562cd7ee03e7f8188a1cfd942858cded/errors.go#L29) That we use to wrap errors.
* You could define a `Reasoner` interface (yeah, weird name) with `Reason()` method. then just have all your errors implement that method. You then type-check (`err.(Reasoner)`) the error, and if it implements it, use `Reason()` to retrieve the type. You can always return "unknown" if type check fails.

kars7e commented on pkg/event-manager/drivers/k8s_backend.go:
```diff
@@ -158,8 +162,10 @@ func (k *k8sBackend) Deploy(driver *entities.Driver) error {
-			Err: ewrapper.Wrapf(err, "k8s: error creating a deployment"),
+		if k8serrors.IsAlreadyExists(err) {
+			err = &EventdriverErrorDeploymentAlreadyExists{
+				Err: ewrapper.Wrapf(err, "k8s: error creating a deployment"),
```
This wrapping seems a bit redundant because `EventdriverErrorDeploymentAlreadyExists` carries context already. Wrapping makes sense if we don't want to deal with declaring custom error types.

kars7e commented on pkg/event-manager/drivers/entity_handler.go:
```diff
@@ -100,13 +106,142 @@ func (h *EntityHandler) Delete(obj entitystore.Entity) error {
+		return
+	}
+	log.Debugf("put driver to error state %s", driver.GetName())
+	reason := []string{
```
Since this is mostly an information to the user, do we need return both `ErrorTypeToErrorReason` and underlying error message? It seems to me that the former would be informative enough.

kars7e commented on pkg/event-manager/drivers/entity_handler.go:
```diff
@@ -100,13 +106,142 @@ func (h *EntityHandler) Delete(obj entitystore.Entity) error {
+	recover := false
+	var err error
+	// try to recover/handle error state
+	switch driver.GetReason()[0] {
```
This can bite you, what if `reason` slice is empty? Seems like "reason" being a `[]string` is problematic. We might need to change that/define another attribute that carries a struct instead. for now, you should at least have a check if `len(reason) > 0`.

kars7e commented on pkg/event-manager/drivers/entity_handler.go:
```diff
@@ -100,13 +106,142 @@ func (h *EntityHandler) Delete(obj entitystore.Entity) error {
+			Subject: "Status",
+			Verb:    entitystore.FilterVerbIn,
+			Object: []entitystore.Status{
+				entitystore.StatusCREATING, entitystore.StatusUPDATING, entitystore.StatusDELETING,
```
Seems like you have all the statuses in this filter, is this filter needed?

kars7e commented on pkg/event-manager/drivers/entity_handler.go:
```diff
@@ -100,13 +106,142 @@ func (h *EntityHandler) Delete(obj entitystore.Entity) error {
+		driver := e.(*entities.Driver)
+
+		switch e.GetStatus() {
+		case entitystore.StatusREADY:
```
`handler.Update()` [is called](https://github.com/vmware/dispatch/blob/186129d49e735fbd5994ba995faded7bc343497b/pkg/controller/controller.go#L142) for entites with READY status, it seems it would make sense for this logic to be moved there?

kars7e commented on pkg/event-manager/drivers/entity_handler.go:
```diff
@@ -100,13 +106,142 @@ func (h *EntityHandler) Delete(obj entitystore.Entity) error {
+	}
+
+	// sync with k8s, pending entities might change status during sync
+	k := h.backend.(*k8sBackend)
```
Let's avoid leaking the implementation of backend. What if we wanted to use different (non-k8s) backend here? I think we could use `Update()` method of `Backend` interface to implement the logic of reconciliation (we could extend that interface as well)

kars7e commented on pkg/event-manager/drivers/entity_handler.go:
```diff
@@ -100,13 +106,142 @@ func (h *EntityHandler) Delete(obj entitystore.Entity) error {
+					}
+				}
+			} else if deployment.Status.AvailableReplicas == 0 {
+				k8serr = &EventdriverErrorDeploymentNotAvaialble{
```
Currently, we set READY state right after deployment is accepted by kubernetes, which means there will be a period of time when available replicas will be 0. This may lead to a weird READY -> ERROR -> READY state flapping, which may be misleading to the user. I guess the solution to that is to ensure there is a replica available before switching from CREATING to READY.

zimengyang committed:
Address comments

* add backoff for k8s Deploy
* move sync logic to k8s backend Update
* remove Error handling, not robust
* refactor EventdriverError types

kars7e committed:
Lift API model structures to be first-class citizens (#434)

* moves API models to a dedicated, common package versioned
in similar fashion as kubernetes API structures
* swagger model definitions are now generated from code
(scripts/generate-models.sh)
* all services use the same model definitions (many have been merged)
* scripts/generate.sh now takes that into account and uses existing
models to generate API handlers stubs
* updated all references to models to use new package location

zimengyang committed:
Add sync and pull reason from Pods

zimengyang commented:
Close PR, separate one on main repo
